### ISSUE TITLE
Openldap plugin fails to bind with Empty result. Check bind username/password.

### ISSUE THREAD

**Issue by scatenag** [netdata/netdata#9004]
##### Question summary
Openldap plugin fails to bind even if ldapsearch works fine with same dn and password.
**Empty result. Check bind username/password. Message:  {'desc': u'No such object'}**
How to debug used dn / fix it ?

##### OS / Environment
Ubuntu 18.04.4 LTS
$OpenLDAP: slapd  (Ubuntu) (May  1 2020 17:11:02)
netadata v1.22.0-12-nightly
Python 2.7.17

**Setup of openldap netdata plugin** with

```
sudo apt install python-ldap
cd /etc/netdata/
sudo ./edit-config python.d/openldap.conf

...
openldap:
 name : 'local'
 username : "cn=monitor,dc=it"   # The bind user with right to access monitor statistics
 password : "MYPASS"                     # The password for the binded user
 server   : 'MYLDAPSERVER'                    # The listening address of the LDAP server. In case of TLS, use the h$
 port     : 389                            # The listening port of the LDAP server. Change to 636 port in case of TLS $
 use_tls  : False                        # Make True if a TLS connection is used
 cert_check : False                         # False if you want to ignore certificate check
 timeout  : 1                              # Seconds to timeout if no connection exit
```

**Ldapsearch works fine**
```
LDAPTLS_REQCERT=never ldapsearch -x -H ldap://MYLDAPSERVER -ZZ -D cn=monitor,dc=it -w MYPASS -b cn=Uptime,cn=Time,cn=Monitor -s base '(objectClass=*)' '*' '+'
..
# search result
search: 3
result: 0 Success
..
```

**Openldap plugin fail to bind**
 
```
sudo /usr/libexec/netdata/plugins.d/python.d.plugin openldap debug trace 
2020-05-12 22:45:03: python.d INFO: plugin[main] : using python v2
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : looking for 'python.d.conf' in ['/etc/netdata', '/usr/lib/netdata/conf.d']
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : loading '/usr/lib/netdata/conf.d/python.d.conf'
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : '/usr/lib/netdata/conf.d/python.d.conf' is loaded
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : looking for 'pythond-jobs-statuses.json' in /var/lib/netdata
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : loading '/var/lib/netdata/pythond-jobs-statuses.json'
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : '/var/lib/netdata/pythond-jobs-statuses.json' is loaded
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : [openldap] looking for 'openldap.conf' in ['/etc/netdata/python.d', '/usr/lib/netdata/conf.d/python.d']
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : [openldap] loading '/etc/netdata/python.d/openldap.conf'
2020-05-12 22:45:03: python.d DEBUG: plugin[main] : [openldap] '/etc/netdata/python.d/openldap.conf' is loaded
2020-05-12 22:45:03: python.d INFO: plugin[main] : [openldap] built 1 job(s) configs
2020-05-12 22:45:03: python.d ERROR: openldap[openldap] : Empty result. Check bind username/password. Message:  {'desc': u'No such object'}
2020-05-12 22:45:03: python.d ERROR: openldap[openldap] : Traceback (most recent call last):
  File "/usr/libexec/netdata/python.d/openldap.chart.py", line 194, in get_data
    result_type, result_data = self.conn.result(num, 1)
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 740, in result
    resp_type, resp_data, resp_msgid = self.result2(msgid,all,timeout)
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 744, in result2
    resp_type, resp_data, resp_msgid, resp_ctrls = self.result3(msgid,all,timeout)
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 751, in result3
    resp_ctrl_classes=resp_ctrl_classes
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 758, in result4
    ldap_result = self._ldap_call(self._l.result4,msgid,all,timeout,add_ctrls,add_intermediates,add_extop)
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 331, in _ldap_call
    reraise(exc_type, exc_value, exc_traceback)
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 315, in _ldap_call
    result = func(*args,**kwargs)
NO_SUCH_OBJECT: {'desc': u'No such object'}

2020-05-12 22:45:03: python.d INFO: plugin[main] : openldap[openldap] : check failed
2020-05-12 22:45:03: python.d INFO: plugin[main] : no jobs to serve
2020-05-12 22:45:03: python.d INFO: plugin[main] : exiting from main...
```

##### Component Name
 netdata python.d.plugin configuration for openldap

Thanks

**Comment by ilyam8** [comment]
@ekartsonakis could you help?

**Comment by scatenag** [comment]
Trying to investigate...

The user seems to be correctly parsed.
The problem seems to be another...

On slapd server, I read

for a successful ldapsearch

```
ber_scanf fmt ({miiiib) ber:
5ecd25e3 >>> dnPrettyNormal: <cn=Operations,cn=Monitor>
5ecd25e3 <<< dnPrettyNormal: <cn=Operations,cn=Monitor>, <cn=operations,cn=monitor>
ber_scanf fmt (m) ber:
ber_scanf fmt ({M}}) ber:
5ecd25e3 ==> limits_get: conn=1000 op=2 self="cn=monitor,dc=it" this="cn=operations,cn=monitor"
5ecd25e3 => monitor_back_search
```

while on python.d.plugin openldap debug trace
```
ber_scanf fmt ({miiiib) ber:
5ecd2600 >>> dnPrettyNormal: <cn=Operations,cn=Monitor>
5ecd2600 <<< dnPrettyNormal: <cn=Operations,cn=Monitor>, <cn=operations,cn=monitor>
ber_scanf fmt (m) ber:
ber_scanf fmt ({M}}) ber:
5ecd2600 ==> limits_get: conn=1000 op=1 self="[anonymous]" this="cn=operations,cn=monitor"
5ecd2600 => monitor_back_search
```

Could be the self="[anonymous]" vs self="cn=monitor,dc=it"  the problem ?

@ekartsonakis: have you any thoughts? Want you a complete slapd log?

Thanks

**Comment by ilyam8** [comment]
@scatenag so what function failes? `simple_bind`?

https://github.com/netdata/netdata/blob/289fb07015c42f676b4b662500827cd38549ff4c/collectors/python.d.plugin/openldap/openldap.chart.py#L164-L165

I dont really can help here, if you have time/willing it would be nice you try to debug/fix it

**Comment by scatenag** [comment]
Okay, I should have been able to fix it.

The problem was that, since [`ldaps:// `is deprecated](https://www.openldap.org/faq/data/cache/605.html), I use TSL over `ldap://` with `start_tsl`;
that case is not covered by the code.

I added the option `use_start_tls` by `sudo ./edit-config python.d/openldap.conf` and managed them in  `/usr/libexec/netdata/python.d/openldap.chart.py` in [a PR](https://github.com/netdata/netdata/pull/9853).

**Comment by ilyam8** [comment]
I beleive this is fixed in #9853

**Comment by ilyam8** [comment]
good work @scatenag
